### ISSUE TITLE
upgrade to 0.20.1 fails

### ISSUE THREAD

**Issue by phynias** [EngineerBetter/concourse-up#98]
i get this part way through

```
....
Task 705 done
  vm_types:
  - name: concourse-large
    cloud_properties:
-     spot_bid_price: 0.13
-     spot_ondemand_fallback: true
  - name: concourse-xlarge
    cloud_properties:
-     spot_bid_price: 0.27
-     spot_ondemand_fallback: true
  - name: concourse-2xlarge
    cloud_properties:
-     spot_bid_price: 0.53
-     spot_ondemand_fallback: true
  - name: concourse-4xlarge
    cloud_properties:
-     spot_bid_price: 1.07
-     spot_ondemand_fallback: true
  - name: concourse-10xlarge
    cloud_properties:
-     spot_bid_price: 2.67
-     spot_ondemand_fallback: true
+ - cloud_properties:
+     ephemeral_disk:
+       encrypted: true
+       size: 200000
+       type: gp2
+     instance_type: m5.12xlarge
+     security_groups:
+     - sg-00ec7cb724c8143b8
+   name: concourse-12xlarge
  - name: concourse-16xlarge
    cloud_properties:
-     spot_bid_price: 4.26
-     spot_ondemand_fallback: true
+ - cloud_properties:
+     ephemeral_disk:
+       encrypted: true
+       size: 200000
+       type: gp2
+     instance_type: m5.24xlarge
+     security_groups:
+     - sg-00ec7cb724c8143b8
+   name: concourse-24xlarge
  - name: compilation
    cloud_properties:
-     ephemeral_disk:
-       encrypted: true
-       size: 5000
-       type: gp2
-     security_groups:
-     - sg-00ec7cb724c8143b8
-     spot_bid_price: 0.13
-     spot_ondemand_fallback: true

  networks:
  - name: public
    subnets:
    - range: 10.0.0.0/24
-     dns:
-     - 10.0.0.2
  - name: private
    subnets:
    - range: 10.0.1.0/24
-     dns:
-     - 10.0.0.2

  stemcells:
+ - alias: xenial
+   os: ubuntu-xenial
+   version: '170.30'
- - alias: xenial
-   os: ubuntu-xenial
-   version: '97.5'

  releases:
  - name: concourse
-   sha1: dbdbb4dba48d93c073130a415b4af307816badfd
+   sha1: bf0b113a8aa48fb063c6cd39cde6107526d02517
    stemcell:
-     version: '97.5'
+     version: '170.30'
-   url: https://s3-eu-west-1.amazonaws.com/concourse-up-compiled-releases/concourse.tgz?versionId=vt5ZifmUd4HmsqNYqxFalHcTu5C5W05Y
+   url: https://s3-eu-west-1.amazonaws.com/concourse-up-compiled-releases-new/concourse-4.2.3-ubuntu-xenial-170.30-20190225-124109-642484759.tgz
-   version: 4.0.0
+   version: 4.2.3
  - name: garden-runc
-   sha1: e89b672e4f48a277cda57537c3ddd00b273a076f
+   sha1: 896f1b94ecbbc117b0a42ac0c889bcd7a7256429
    stemcell:
-     version: '97.5'
+     version: '170.30'
-   url: https://s3-eu-west-1.amazonaws.com/concourse-up-compiled-releases/garden-runc.tgz?versionId=f53fVACEfZs76tvQuPB2GuOptNHNukm8
+   url: https://s3-eu-west-1.amazonaws.com/concourse-up-compiled-releases-new/garden-runc-1.18.3-ubuntu-xenial-170.30-20190225-124157-21139916.tgz
-   version: 1.16.2
+   version: 1.18.3
  - name: riemann
-   sha1: e335be336ece7f83f682faae28a8957bc06bc277
+   sha1: a98ccebb259e446d1ea830efa811e9816536a5dc
    stemcell:
-     version: '97.5'
+     version: '170.30'
-   url: https://s3-eu-west-1.amazonaws.com/concourse-up-compiled-releases/riemann.tgz?versionId=IW2sS.efrBrd5e2bAGLunj.DJ6oMxfY4
+   url: https://s3-eu-west-1.amazonaws.com/concourse-up-compiled-releases-new/riemann-5-ubuntu-xenial-170.30-20190225-124226-242639627.tgz
  - name: grafana
-   sha1: 1a34a62c0e5e338cfac1b779c14852c3458f24ad
+   sha1: 05b4d65ed61428638f0ffcccfd14605cd75b7cb1
    stemcell:
-     version: '97.5'
+     version: '170.30'
-   url: https://s3-eu-west-1.amazonaws.com/concourse-up-compiled-releases/grafana.tgz?versionId=KJvU5vhc3ZqOslSX_ql_kp6oW2cwoaUS
+   url: https://s3-eu-west-1.amazonaws.com/concourse-up-compiled-releases-new/grafana-13.2.0-ubuntu-xenial-170.30-20190225-124242-918670614.tgz
  - name: influxdb
-   sha1: 0ebed50b3a1532ad7b63c198d98da20d48cfb00b
+   sha1: ba2d2271ba8e96cc62cc478b4717dc0717eab5fe
    stemcell:
-     version: '97.5'
+     version: '170.30'
-   url: https://s3-eu-west-1.amazonaws.com/concourse-up-compiled-releases/influxdb.tgz?versionId=sBPXBsOH5yoOYc8C.7Kv6nYKkvHiqk_L
+   url: https://s3-eu-west-1.amazonaws.com/concourse-up-compiled-releases-new/influxdb-4-ubuntu-xenial-170.30-20190225-124254-534283147.tgz
  - name: credhub
-   sha1: b9e61a2f000b530b36a82ef12725dd56f767bb37
+   sha1: 2960f6166b35052d94eb286241a56bede4d6e6f5
    stemcell:
-     version: '97.5'
+     version: '170.30'
-   url: https://s3-eu-west-1.amazonaws.com/concourse-up-compiled-releases/credhub.tgz?versionId=O82NZqcYl2gwXG47Gmfi.w3rYOBdyTJI
+   url: https://s3-eu-west-1.amazonaws.com/concourse-up-compiled-releases-new/credhub-2.1.2-ubuntu-xenial-170.30-20190225-124319-586905766.tgz
-   version: 2.0.2
+   version: 2.1.2
  - name: uaa
-   sha1: 4fb62da4a02148e9ecf7382ddbcd740e46a4c2f3
+   sha1: 8436c40c26302d00f3a05c8f320a93f81ee271d8
    stemcell:
-     version: '97.5'
+     version: '170.30'
-   url: https://s3-eu-west-1.amazonaws.com/concourse-up-compiled-releases/uaa.tgz?versionId=kR.R23UoYJrBfsR3XTr4Nfc.WkK1WZPB
+   url: https://s3-eu-west-1.amazonaws.com/concourse-up-compiled-releases-new/uaa-70.0-ubuntu-xenial-170.30-20190225-124400-918123017.tgz
-   version: '60.2'
+   version: '70.0'
+ - name: bosh-dns
+   sha1: 73714da658e1966a0b88f076aeb25fdf1911360d
+   stemcell:
+     os: ubuntu-xenial
+     version: '170.30'
+   url: https://s3-eu-west-1.amazonaws.com/concourse-up-compiled-releases-new/bosh-dns-1.10.0-ubuntu-xenial-170.30-20190225-124433-113857975.tgz
+   version: 1.10.0

  update:
-   canary_watch_time: 1000-300000
+   canary_watch_time: 1000-600000
-   update_watch_time: 1000-300000
+   update_watch_time: 1000-600000

  instance_groups:
  - name: web
    jobs:
    - name: atc
      properties:
        influxdb:
-         database: "<redacted>"
+         database: "<redacted>"
-       tls_cert: "<redacted>"
+       tls_cert: "<redacted>"
-       tls_key: "<redacted>"
+       tls_key: "<redacted>"
    - name: grafana
      properties:
        grafana:
          ssl:
-           cert: "<redacted>"
+           cert: "<redacted>"
-           key: "<redacted>"
+           key: "<redacted>"
+   - name: bosh-dns
+     properties:
+       api:
+         client:
+           tls:
+             ca: "<redacted>"
+             certificate: "<redacted>"
+             private_key: "<redacted>"
+         server:
+           tls:
+             ca: "<redacted>"
+             certificate: "<redacted>"
+             private_key: "<redacted>"
+       cache:
+         enabled: "<redacted>"
+     release: bosh-dns
  - name: worker
    jobs:
    - name: garden
      properties:
        garden:
+         no_image_plugin: "<redacted>"

  tags:
+   concourse-up-version: 0.20.1

Task 706

Task 706 | 19:32:10 | Preparing deployment: Preparing deployment (00:00:02)
Task 706 | 19:32:12 | Preparing deployment: Rendering templates (00:00:07)
                    L Error: Unable to render instance groups for deployment. Errors are:
  Blobstore object 'c0e4cc50-ebd3-4b1d-8cfa-7477153f26e9' not found
Task 706 | 19:32:19 | Error: Unable to render instance groups for deployment. Errors are:
  Blobstore object 'c0e4cc50-ebd3-4b1d-8cfa-7477153f26e9' not found

Task 706 Started  Tue Feb 26 19:32:10 UTC 2019
Task 706 Finished Tue Feb 26 19:32:19 UTC 2019
Task 706 Duration 00:00:09
Task 706 error

Updating deployment:
  Expected task '706' to succeed but state is 'error'

Exit code 1
failed to run bosh deploy with commands [/var/folders/nm/f4mbqc4d19s1x9m745_r1vx40000gn/T/concourse-up277427764/concourse.yml --vars-store /var/folders/nm/f4mbqc4d19s1x9m745_r1vx40000gn/T/concourse-up277427764/concourse-creds.yml --ops-file /var/folders/nm/f4mbqc4d19s1x9m745_r1vx40000gn/T/concourse-up277427764/versions.json --ops-file /var/folders/nm/f4mbqc4d19s1x9m745_r1vx40000gn/T/concourse-up277427764/shas.json --ops-file /var/folders/nm/f4mbqc4d19s1x9m745_r1vx40000gn/T/concourse-up277427764/cup_compatibility.yml --vars-file /var/folders/nm/f4mbqc4d19s1x9m745_r1vx40000gn/T/concourse-up277427764/grafana_dashboard.yml --ops-file /var/folders/nm/f4mbqc4d19s1x9m745_r1vx40000gn/T/concourse-up277427764/extra_tags.yml]: [exit status 1]
```

**Comment by crsimmons** [comment]
Hi @phynias. What version of Concourse-up are you trying to upgrade from? The diff output looks like its a big jump.

We have system tests checking upgrades but we only test consecutive releases.

**Comment by phynias** [comment]
not sure what version i was on, and it appears there is no way to tell what version i was coming from. 
so not sure what to do from here.

**Comment by phynias** [comment]
also since i tried to upgrade to a way newer version it upgraded some things already. 
for instance i think i found the version i was on and so i tried to upgrade to the next version. it not complains about the version of terraform being newer and fails. 

if anyone can help me figure out what to do next to get this upgraded and fixed i would appreciate it.
also, maybe it would be a good idea to prevent an upgrade from even happening if the versions are not consecutive?

**Comment by crsimmons** [comment]
Based on the versions you had before in the diff you were coming from either 0.11.1 or 0.11.2 which are both from August last year. I would generally expect non-consecutive releases to work fine with upgrades (although we don't explicitly test it). We've made some major refactors since August though and pretty much all of the dependencies have bumped substantially. Its an interesting idea to put some kind of version delta limit in place though.

I'll try replicating your upgrade from 0.11.1 -> 0.20.2 tomorrow. I haven't seen your particular error before. Can you run `eval "$(concourse-up info --region <your region> --env <your deployment>)"`? That will export the env vars needed to interact with bosh and credhub. If that works then you can poke around with [bosh](https://github.com/cloudfoundry/bosh-cli/releases) to get a better feeling of what is going wrong. You could start with `bosh vms` to see what vms are failing `bosh cck` may also be able to get you out of a mess.

The simplest solution may actually be to just cut your losses and deploy a new concourse if this is acceptable given your circumstances. If that eval command works you should be able to export all the credhub secrets with `credhub export -f filename.yml` then import them to a new concourse with `credhub import -f filename.yml`. Then just set all the pipelines/teams again. The only thing you should lose is build history.

**Comment by DanielJonesEB** [comment]
FWIW @phynias the team are working on an export/import tool, but I wouldn't wait for it.

**Comment by phynias** [comment]
```
$ bosh vms
Using environment '13.59.59.30' as client 'admin'

Task 736. Done

Deployment 'concourse'

Instance                                     Process State  AZ  IPs           VM CID               VM Type              Active
web/7a43297a-38b1-4178-a7f0-e53322cc362a     running        z1  10.0.0.8      i-0fb9000e08a026e2a  concourse-web-small  true
                                                                18.219.24.59
worker/60ad6509-a4ea-46a5-8742-1a8d9b8a3f2f  running        z1  10.0.1.8      i-09ca0986d16dd355b  concourse-xlarge     true
worker/bc9ad1a1-d8c9-4b4a-a0d5-d63dfdcad5f6  running        z1  10.0.1.6      i-0585ef4c04bc1c929  concourse-xlarge     true
worker/e577709a-f017-415f-af0e-afa082375d1f  running        z1  10.0.1.7      i-06b6591c251347f2f  concourse-xlarge     true

4 vms

Succeeded
```


 ```
$ bosh cck
Using environment '13.59.59.30' as client 'admin'

Using deployment 'concourse'

Task 737

Task 737 | 19:51:17 | Scanning 4 VMs: Checking VM states (00:00:12)
Task 737 | 19:51:29 | Scanning 4 VMs: 4 OK, 0 unresponsive, 0 missing, 0 unbound (00:00:00)
Task 737 | 19:51:29 | Scanning 0 persistent disks: Looking for inactive disks (00:00:00)
Task 737 | 19:51:29 | Scanning 0 persistent disks: 0 OK, 0 missing, 0 inactive, 0 mount-info mismatch (00:00:00)

Task 737 Started  Mon Mar  4 19:51:17 UTC 2019
Task 737 Finished Mon Mar  4 19:51:29 UTC 2019
Task 737 Duration 00:00:12
Task 737 done

#  Type  Description

0 problems

Succeeded
```

**Comment by phynias** [comment]
yeah. looks like i may have to destroy and rebuild it.

but it may be a good idea to put something in place to prevent the upgrade from even starting if the versions are too big. my fault for not looking but i honestly didn't even think about it.

thanks for the help!

**Comment by DanielJonesEB** [comment]
It's curious that it's an error about a missing blobstore object.

@phynias has something happened to your blobstore bucket?

**Comment by crsimmons** [comment]
It looks like bosh still thinks everything is fine. You might be able to get a running Concourse with a deploy. `bosh deploy <(bosh manifest)`

**Comment by phynias** [comment]
so let me clarify. nothing is broken. concourse is still running fine. i just can't upgrade it.

**Comment by DanielJonesEB** [comment]
BOSH is complaining that it can't find something in its blobstore. This is different to a general BOSH blobstore where final releases are published for the general public to use - it's the S3 bucket with `blobstore` and the name of your Concourse-Up deployment in it. BOSH will unpack releases (Concourse, CredHub, and others) into this blobstore bucket.

BOSH is erring when trying to render templates; presumably it's missing a blob that contains ERB templates for some of the job config. This suggests that the bucket has gone, or has been emptied. As @crsimmons suggests, doing something like `bosh manifest > manifest.yml` then `bosh deploy -d concourse manifest.yml` might prompt the BOSH Director to rebuild those blobs from the compiled releases we provide. Once that's done, then it might unblock you from upgrading.

It's the evening here in the UK, but we can look into this more in our working day tomorrow to check our assumptions.

**Comment by phynias** [comment]
i tried the bosh deploy and pretty much the same errors.
I will wait to see if you guys find anything tomorrow if not i will just destroy and redeploy.

thanks!

**Comment by DanielJonesEB** [comment]
@phynias What is the state of your blobstore bucket? Has it been deleted? Have any files been deleted from it?

**Comment by crsimmons** [comment]
I've done some investigation and have replicated your issue. Essentially, somewhere between versions 0.11.1 and 0.20.1 there's a bug where namespace (introduced in 0.13.7) [doesn't get templated properly into the terraform](https://github.com/EngineerBetter/concourse-up/blob/master/resource/assets/aws/infrastructure.tf#L110). This works if you upgrade incrementally but this big bang jump breaks things for some reason. The result is that terraform thinks the blobstore bucket name needs to be renamed from `concourse-up-yourdeployment-region-blobstore` to `concourse-up-yourdeployment--blobstore`. When terraform renames a bucket it actually deletes the old one and everything in it before making the new one. So your old blobstore got destroyed. IMO this is an edge case that we probably won't investigate further.

I did manage to get out of the broken state though, and thus was able to finish the upgrade.

TL;DR: run the following against your BOSH director then run `concourse-up deploy` again with the 0.20.1 binary:
```sh
bosh upload-release https://s3-eu-west-1.amazonaws.com/concourse-up-compiled-releases-new/concourse-4.2.3-ubuntu-xenial-170.30-20190225-124109-642484759.tgz --version "4.2.3" --sha1 "bf0b113a8aa48fb063c6cd39cde6107526d02517" --fix
bosh upload-release https://s3-eu-west-1.amazonaws.com/concourse-up-compiled-releases-new/garden-runc-1.18.3-ubuntu-xenial-170.30-20190225-124157-21139916.tgz --version "1.18.3" --sha1 "896f1b94ecbbc117b0a42ac0c889bcd7a7256429" --fix
bosh upload-release https://s3-eu-west-1.amazonaws.com/concourse-up-compiled-releases-new/riemann-5-ubuntu-xenial-170.30-20190225-124226-242639627.tgz --version "5" --sha1 "a98ccebb259e446d1ea830efa811e9816536a5dc" --fix
bosh upload-release https://s3-eu-west-1.amazonaws.com/concourse-up-compiled-releases-new/grafana-13.2.0-ubuntu-xenial-170.30-20190225-124242-918670614.tgz --version "13.2.0" --sha1 "05b4d65ed61428638f0ffcccfd14605cd75b7cb1" --fix
bosh upload-release https://s3-eu-west-1.amazonaws.com/concourse-up-compiled-releases-new/influxdb-4-ubuntu-xenial-170.30-20190225-124254-534283147.tgz --version "4" --sha1 "ba2d2271ba8e96cc62cc478b4717dc0717eab5fe" --fix
bosh upload-release https://s3-eu-west-1.amazonaws.com/concourse-up-compiled-releases-new/credhub-2.1.2-ubuntu-xenial-170.30-20190225-124319-586905766.tgz --version "2.1.2" --sha1 "2960f6166b35052d94eb286241a56bede4d6e6f5" --fix
bosh upload-release https://s3-eu-west-1.amazonaws.com/concourse-up-compiled-releases-new/uaa-70.0-ubuntu-xenial-170.30-20190225-124400-918123017.tgz --version "70.0" --sha1 "8436c40c26302d00f3a05c8f320a93f81ee271d8" --fix
bosh upload-release https://s3-eu-west-1.amazonaws.com/concourse-up-compiled-releases-new/bosh-dns-1.10.0-ubuntu-xenial-170.30-20190225-124433-113857975.tgz --version "1.10.0" --sha1 "73714da658e1966a0b88f076aeb25fdf1911360d" --fix
```
Then I was even able to upgrade to 0.20.2 successfully! The blobstore bucket name will remain a little weird (missing the region/namespace) though so if you've got the time it might still be a good idea to deploy a new one.

Essentially you need to reupload all the releases that got deleted from the blobstore. The `--fix` flag tells the director to overwrite releases with the same name. I got the urls and shas from the version of [concourse-up-ops](https://github.com/EngineerBetter/concourse-up-ops/tree/0.0.174) used to build 0.20.1 of Concourse-up. (see `ops/versions-aws.json` and `ops/shas-aws.json` in that repo).

**Comment by phynias** [comment]
if there is an edge case i will find it.
thanks for looking into this. i think i will redeploy though, i'd rather have a clean install.
